src/tiled/mapdocument.cpp

```diff
@@ -586,6 +586,8 @@ void MapDocument::moveLayerDown(Layer *layer)
 void MapDocument::removeLayer(Layer *layer)
 {
     Q_ASSERT(layer->map() == mMap);
+    if (!layer->isUnlocked())
+        return;
```

Comment:
You should be able to remove locked layers. Also this check would be in the wrong place since this function is not expected to fail.